server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingMetadata.java

```diff
@@ -0,0 +1,315 @@
+            out.writeEnum(this);
+        }
+    }
+
+    public enum TargetShardState implements Writeable {
+        CLONE,
+        HANDOFF,
+        SPLIT,
+        DONE;
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeEnum(this);
+        }
+    }
+
+    // Copying from IndexMetadataStats here
+    public static final ParseField OLD_SHARD_COUNT_FIELD = new ParseField("old_shard_count");
+    public static final ParseField NEW_SHARD_COUNT_FIELD = new ParseField("new_shard_count");
+    public static final ParseField SOURCE_SHARD_STATES_FIELD = new ParseField("source_shard_states");
+    public static final ParseField TARGET_SHARD_STATES_FIELD = new ParseField("target_shard_states");
+
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<IndexReshardingMetadata, Void> PARSER = new ConstructingObjectParser<>(
+        "index_resharding_metadata_parser",
+        false,
+        (args, unused) -> new IndexReshardingMetadata(
+            (int) args[0],
+            (int) args[1],
+            ((List<SourceShardState>) args[2]).toArray(new SourceShardState[0]),
+            ((List<TargetShardState>) args[3]).toArray(new TargetShardState[0])
+        )
+    );
+
+    static {
+        PARSER.declareInt(ConstructingObjectParser.constructorArg(), OLD_SHARD_COUNT_FIELD);
+        PARSER.declareInt(ConstructingObjectParser.constructorArg(), NEW_SHARD_COUNT_FIELD);
+        // XXX I'm not sure this is the best way to parse an array of enums
+        PARSER.declareObjectArray(
+            ConstructingObjectParser.constructorArg(),
+            (parser, c) -> SourceShardState.valueOf(parser.text()),
+            SOURCE_SHARD_STATES_FIELD
+        );
+        PARSER.declareObjectArray(
+            ConstructingObjectParser.constructorArg(),
+            (parser, c) -> TargetShardState.valueOf(parser.text()),
+            TARGET_SHARD_STATES_FIELD
+        );
+    }
+
+    static IndexReshardingMetadata fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, null);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field(OLD_SHARD_COUNT_FIELD.getPreferredName(), oldShardCount);
+        builder.field(NEW_SHARD_COUNT_FIELD.getPreferredName(), newShardCount);
+        builder.field(SOURCE_SHARD_STATES_FIELD.getPreferredName(), sourceShardStates);
+        builder.field(TARGET_SHARD_STATES_FIELD.getPreferredName(), targetShardStates);
+        return builder;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeInt(oldShardCount);
+        out.writeInt(newShardCount);
+        out.writeArray(sourceShardStates);
+        out.writeArray(targetShardStates);
+    }
+
+    public IndexReshardingMetadata(StreamInput in) throws IOException {
+        this(
+            in.readInt(),
```

Comment:
Not that it matters that much, but I don't think we need to store this if we are also storing an array of sources and targets. We can just read the array size.

Reply:
That's true. An earlier revision wasn't tracking source shard state, so I needed it. I'm kind of tempted to keep it just for clarity, but I agree it's redundant.

Reply:
In my opinion it is fine to keep it as a dedicated field in memory, however, we should not serialize additional data in the cluster state that is not needed.

Reply:
ok

Reply:
Done in 7f6c7b9.